resize.c

```diff
@@ -42,50 +43,95 @@
 #endif
 #endif
 
-/**
- * mutt_resize_screen - Called after receiving SIGWINCH
- */
-void mutt_resize_screen(void)
+int mutt_resize_getrows(int rows)
 {
   const char *cp = NULL;
-  int fd;
-  struct winsize w;
-#if !defined(USE_SLANG_CURSES)
-  int SLtt_Screen_Rows, SLtt_Screen_Cols;
-#endif
+  cp = mutt_str_getenv("LINES");
+  if (cp && (mutt_str_atoi(cp, &rows) < 0))
+    return 24;
+  return -1;
+}
 
-  SLtt_Screen_Rows = -1;
-  SLtt_Screen_Cols = -1;
+int mutt_resize_getcols(int cols)
+{
+  const char *cp = NULL;
+  cp = mutt_str_getenv("COLUMNS");
+  if (cp && (mutt_str_atoi(cp, &cols) < 0))
+    return 80;
+  return -1;
+}
+
+int mutt_resize_getwinsize(winsize w)
+{
+  int fd;
+  int ret = 0;
   fd = open("/dev/tty", O_RDONLY);
   if (fd != -1)
   {
     if (ioctl(fd, TIOCGWINSZ, &w) != -1)
-    {
-      SLtt_Screen_Rows = w.ws_row;
-      SLtt_Screen_Cols = w.ws_col;
-    }
+      ret = 1;
     close(fd);
   }
-  if (SLtt_Screen_Rows <= 0)
-  {
-    cp = mutt_str_getenv("LINES");
-    if (cp && (mutt_str_atoi(cp, &SLtt_Screen_Rows) < 0))
-      SLtt_Screen_Rows = 24;
+  return ret;
+}
+
+#if defined(USE_SLANG_CURSES)
+void mutt_resize_screen_slang(void)
+{
+  struct winsize w;
+
+  /* The following two variables are global to Slang */
+  SLtt_Screen_Rows = -1;
+  SLtt_Screen_Cols = -1;
+
+  if (mutt_resize_getwinsize(w)) {
+    SLtt_Screen_Rows = w.ws_row;
+    SLtt_Screen_Cols = w.ws_col;
   }
+
+  if (SLtt_Screen_Rows <= 0)
+    SLtt_Screen_Rows = mutt_resize_getrows(SLtt_Screen_Rows);
+
   if (SLtt_Screen_Cols <= 0)
-  {
-    cp = mutt_str_getenv("COLUMNS");
-    if (cp && (mutt_str_atoi(cp, &SLtt_Screen_Cols) < 0))
-      SLtt_Screen_Cols = 80;
-  }
-#ifdef USE_SLANG_CURSES
+    SLtt_Screen_Cols = mutt_resize_getcols(SLtt_Screen_Cols);
+
   delwin(stdscr);
   SLsmg_reset_smg();
   SLsmg_init_smg();
   stdscr = newwin(0, 0, 0, 0);
   keypad(stdscr, true);
+  mutt_window_reflow();
+}
 #else
-  resizeterm(SLtt_Screen_Rows, SLtt_Screen_Cols);
-#endif
+void mutt_resize_screen_ncurses(void)
+{
+  struct winsize w;
+  int screenrows = -1, screencols = -1;
+
+  if (mutt_resize_getwinsize(w)) {
+    screenrows = w.ws_row;
+    screencols = w.ws_col;
+  }
+
+  if (screenrows <= 0)
+    screenrows = mutt_resize_getrows(screenrows);
+
+  if (screencols <= 0)
+    screencols = mutt_resize_getcols(screencols);
+
+  resizeterm(screenrows, screencols);
   mutt_window_reflow();
 }
+#endif
+
+/**
+ * mutt_resize_screen - Called after receiving SIGWINCH
+ */
+void mutt_resize_screen(void)
```

Comment:
I don't think we really need this extra level of abstraction.
We can simply rename the two functions to `mutt_resize_screen()` and `#ifdef` them.